internal/configs/virtualserver.go

```diff
@@ -490,169 +546,212 @@ type policiesCfg struct {
 	ErrorReturn     *version2.Return
 }
 
-// TODO refactor generatePolicies
-func (vsc *virtualServerConfigurator) generatePolicies(owner runtime.Object, ownerNamespace string, vsNamespace string,
-	vsName string, policyRefs []conf_v1.PolicyReference, policies map[string]*conf_v1alpha1.Policy, jwtKeys map[string]string, ingressMTLSPemFileName string, context string, tlsPemFileName string, egressMTLSSecrets map[string]string) policiesCfg {
-	var policyErrorReturn *version2.Return
-	var allow, deny []string
-	var limitReqOptions version2.LimitReqOptions
-	var limitReqZones []version2.LimitReqZone
-	var limitReqs []version2.LimitReq
-	var JWTAuth *version2.JWTAuth
-	var ingressMTLS *version2.IngressMTLS
-	var egressMTLS *version2.EgressMTLS
-	var policyError bool
+func newPoliciesConfig() *policiesCfg {
+	return &policiesCfg{}
+}
 
-	for _, p := range policyRefs {
-		polNamespace := p.Namespace
-		if polNamespace == "" {
-			polNamespace = ownerNamespace
+type policyOwnerDetails struct {
+	owner          runtime.Object
+	ownerNamespace string
+	vsNamespace    string
+	vsName         string
+}
+
+type policyOptions struct {
+	jwtKeys                map[string]string
+	ingressMTLSPemFileName string
+	tlsPemFileName         string
+	egressMTLSSecrets      map[string]string
+}
+
+type validationResults struct {
+	isError  bool
+	warnings []string
+}
+
+func newValidationResults() *validationResults {
+	return &validationResults{}
+}
+
+func (v *validationResults) addWarningf(msgFmt string, args ...interface{}) {
+	v.warnings = append(v.warnings, fmt.Sprintf(msgFmt, args...))
+}
+
+func (p *policiesCfg) addAccessControlConfig(accessControl *conf_v1alpha1.AccessControl) *validationResults {
+	res := newValidationResults()
+	p.Allow = append(p.Allow, accessControl.Allow...)
+	p.Deny = append(p.Deny, accessControl.Deny...)
+	if len(p.Allow) > 0 && len(p.Deny) > 0 {
+		res.addWarningf("AccessControl policy (or policies) with deny rules is overridden by policy (or policies) with allow rules")
+	}
+	return res
+}
+
+func (p *policiesCfg) addRateLimitConfig(rateLimit *conf_v1alpha1.RateLimit, polKey string, polNamespace string, polName string, vsNamespace string, vsName string) *validationResults {
```

Comment:
Optional suggestion:
* Reduce line length

```suggestion
func (p *policiesCfg) addRateLimitConfig(rateLimit *conf_v1alpha1.RateLimit, polKey string, polNs string, polName string, vsNs string, vsName string) *validationResults {
```
**Note**: Probably makes sense to also do for `addJWTAuthConfig` if you do this change.

Or
```suggestion
func (p *policiesCfg) addRateLimitConfig(rateLimit *conf_v1alpha1.RateLimit, polKey string, polNamespace string,
	polName string, vsNamespace string, vsName string) *validationResults {
```